File: src/test/java/org/hejwo/java/algorithms/sorting/mergesort/MergeSortTest.java
package org.hejwo.java.algorithms.sorting.mergesort;

import java.util.List;
import java.util.StringJoiner;

import org.apache.commons.lang3.tuple.Pair;
import org.junit.Test;

import com.google.common.collect.Lists;

import static org.assertj.core.api.Assertions.assertThat;

public class MergeSortTest {

    @Test
    public void mergeResults() {
        List<Integer> result1 = Lists.newArrayList(4, 5, 6);
        List<Integer> result2 = Lists.newArrayList(1, 2, 3);

        List<Integer> merged = MergeSort.mergeResults(Pair.of(result1, result2));
        printResult(merged);

        assertThat(merged).containsExactly(1, 2, 3, 4, 5, 6);
    }

    @Test
    public void mergeSort() {
        List<Integer> unsorted = Lists.newArrayList(32, 1, 2, 4, 3, 6, 7, 8, 5);

        List<Integer> sorted = MergeSort.sort(unsorted);
        printResult(sorted);

        assertThat(sorted).containsExactly(1, 2, 3, 4, 5, 6, 7, 8, 32);
    }

    @Test
    public void sort_shouldHandleSimpleCase() {
        List<Integer> unsorted = Lists.newArrayList(8, 7, 1, 2, 4, 3, 6, 5);

        List<Integer> sorted = MergeSort.sort(unsorted);

        assertThat(sorted).containsExactly(1, 2, 3, 4, 5, 6, 7, 8);
    }

    @Test
    public void sort_shouldReturnSorted_whenAlreadySorted() {
        List<Integer> unsorted = Lists.newArrayList(1, 2, 3, 4, 5, 6, 7, 8);

        List<Integer> sorted = MergeSort.sort(unsorted);

        assertThat(sorted).containsExactly(1, 2, 3, 4, 5, 6, 7, 8);
    }

    @Test
    public void sort_shouldReturnSameOne_whenEmpty() {
        List<Integer> unsorted = Lists.newArrayList();

        List<Integer> sorted = MergeSort.sort(unsorted);

        assertThat(sorted).isEqualTo(unsorted);
    }

    @Test
    public void sort_shouldReturnSameOne_whenSingleElement() {
        List<Integer> unsorted = Lists.newArrayList(124);

        List<Integer> sorted = MergeSort.sort(unsorted);

        assertThat(sorted).isEqualTo(unsorted);
    }

    @Test
    public void sort_shouldReturnSameOne_whenNull() {
        List<Integer> unsorted = null;

        List<Integer> sorted = MergeSort.sort(unsorted);

        assertThat(sorted).isEqualTo(unsorted);
    }    

    private void printResult(List<Integer> sorted) {
        StringJoiner stringJoiner = new StringJoiner(",");
        sorted.forEach(val -> stringJoiner.add(val.toString()));
        System.out.println(stringJoiner.toString());
    }

}

File: src/test/scala/org/hejwo/scala/algorithms/sorting/mergesort/MergeSortTest.scala
package org.hejwo.scala.algorithms.sorting.mergesort

import org.scalatest.{FlatSpec, Matchers}

class MergeSortTest extends FlatSpec with Matchers {

  "sort" should "return empty when empty" in {
    val list = List()
    val sorted = MergeSort.sort(list)

    sorted shouldBe empty
  }

  "sort" should "return single element when single element" in {
    val list = List(1)
    val sorted = MergeSort.sort(list)

    sorted should have size 1
  }

  "sort" should "sort simple list" in {
    val list = List(2, 1)
    val sorted = MergeSort.sort(list)

    sorted shouldBe List(1, 2)
  }

  "sort" should "sort with duplicates" in {
    val unsorted = List(3, 1, 4, 2, 4, 5)
    val sorted = MergeSort.sort(unsorted)

    sorted shouldBe List(1, 2, 3, 4, 4, 5)
  }

  "merge" should "join for right array smaller and equal size" in {
    val list1 = List(4, 5, 6)
    val list2 = List(1, 2, 3)

    val merged = MergeSort.merge(list1, list2)

    merged shouldBe List(1, 2, 3, 4, 5, 6)
  }

  "merge" should "join for right array smaller" in {
    val list1 = List(4, 5, 6, 7)
    val list2 = List(1, 2)

    val merged = MergeSort.merge(list1, list2)

    merged shouldBe List(1, 2, 4, 5, 6, 7)
  }

  "mergeRec" should "join for right array smaller and equal size" in {
    val list1 = List(4, 5, 6)
    val list2 = List(1, 2, 3)

    val merged = MergeSort.mergeRec(list1, list2)

    merged shouldBe List(1, 2, 3, 4, 5, 6)
  }

  "mergeRec" should "join for right array smaller" in {
    val list1 = List(4, 5, 6, 7)
    val list2 = List(1, 2)

    val merged = MergeSort.mergeRec(list1, list2)

    merged shouldBe List(1, 2, 4, 5, 6, 7)
  }

}
